MIDAS/src/hardware/GPSSensor.cpp

```diff
@@ -17,17 +17,53 @@ ErrorCode GPSSensor::init() {
     return ErrorCode::NoError;
 }
 
+// This is needed because GPS doesn't provide unix time and just gives
+// dd mm yy
+uint16_t months[12] = {
+    0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334
+};
+
+inline bool is_leapyear(int year) {
+    return (!(year % 100 == 0) || (year % 400 == 0)) && year % 4 == 0;
+}
+
 GPS GPSSensor::read() {
     teseo.update();
     GPGGA_Info_t gpgga_message = teseo.getGPGGAData();
     GPRMC_Info_t gprmc_message = teseo.getGPRMCData();
-    // Divide latitude and longitude by 100 because NMEA data is ddmmm.mmm or something like that
-    // Also check if it's north or south
-    float lat = gpgga_message.xyz.lat / 100.f * ((gpgga_message.xyz.ns == 'N') ? 1 : -1);
-    float lon = gpgga_message.xyz.lon / 100.f * ((gpgga_message.xyz.ns == 'E') ? 1 : -1);
+
+    float64_t lat = gpgga_message.xyz.lat;
+    float64_t lon = gpgga_message.xyz.lon;
+
+    //                                            d ddm m.mm mmm
+    // the max value of an unsigned 32 bit int is 2,147,4 83,647
+    // Since the maximum longitude is 180, we can store 3 degree digits, and
+    // 7 minute digits, which is all we need, because NMEA gives 6 minute digits.
+    // See https://www.sparkfun.com/datasheets/GPS/NMEA%20Reference%20Manual-Rev2.1-Dec07.pdf
+    int32_t lat_int = static_cast<int32_t>(lat*100000);
+    int32_t lon_int = static_cast<int32_t>(lon*100000);
+
+    lat_int *= (gpgga_message.xyz.ns == 'N') ? 1 : -1;
+    lon_int *= (gpgga_message.xyz.ew == 'E') ? 1 : -1;
     float alt = gpgga_message.xyz.alt;
     float v = gprmc_message.speed;
     uint16_t sat_count = gpgga_message.sats;
 
-    return GPS{lat, lon, alt, v, sat_count};
+    if (!is_leapyear(gprmc_message.date % 100)) {
+        is_leap = true;
+    }
+    if (!is_leap && is_leapyear(2000 + gprmc_message.date % 100)) {
+        is_leap = true;
+        for (int i = 2; i < 12; i++) {
+            months[i]++;
```

Comment:
why are we modifying a global array

Reply:
also this is fragile in the case that the rocket is turned on in december 2024 but is left running until is launched in february 2025